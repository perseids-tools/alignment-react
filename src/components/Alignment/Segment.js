import AlignmentContext from './alignment-context';

import Segment from '../Segment';

const AdditionalFieldsWithContext = () => (
  <AlignmentContext.Consumer>
    {({ json }) => <Segment json={json} />}
  </AlignmentContext.Consumer>
);

export default AdditionalFieldsWithContext;
