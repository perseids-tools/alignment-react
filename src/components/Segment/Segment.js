import React from 'react';
import { func, string } from 'prop-types';

import { activeType, sentenceType } from '../../types';

import styles from './Segment.module.scss';

const wordSpans = (words, lnum, active, toggleActive) => {
  const spans = [];

  words.w.forEach((word) => {
    const { text, $: { n } } = word;
    const classes = [styles.word];
    const innerText = text.map(({ _ }) => _).join(<br />);

    if (active && active[lnum] && active[lnum].has(n)) {
      classes.push(styles.active);
    }
    const onClick = () => {
      toggleActive([lnum, n]);
    };
    const onKeyDown = (event) => {
      const { key } = event;

      if (key === 'Enter') {
        onClick();
      }
    };

    spans.push(
      <span
        key={n}
        role="button"
        tabIndex="0"
        className={classes.join(' ')}
        onClick={onClick}
        onKeyDown={onKeyDown}
      >
        {innerText}
      </span>,
    );
    spans.push(' ');
  });

  spans.pop();

  return spans;
};

const Segment = ({
  lnum, sentence, active, toggleActive,
}) => {
  const words = (sentence && sentence.wds) ? sentence.wds.find(({ $ }) => $.lnum === lnum) : false;

  if (!words) {
    return false;
  }

  return (
    <div className={styles.text}>
      {wordSpans(words, lnum, active, toggleActive)}
    </div>
  );
};

Segment.propTypes = {
  lnum: string.isRequired,
  sentence: sentenceType.isRequired,
  active: activeType,
  toggleActive: func.isRequired,
};

Segment.defaultProps = {
  active: null,
};

export default Segment;
