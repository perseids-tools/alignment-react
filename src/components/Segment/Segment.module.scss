.text {
  background-color: #f2f2f2;
  text-align: left;

  padding-top: 0.2rem;
  padding-bottom: 0.2rem;
  padding-left: 0.6rem;
  padding-right: 0.6rem;

  margin-bottom: 0.2rem;
}

.word {
  cursor: pointer;
  padding-left: 1px;
  padding-right: 1px;
  font-weight: 450;

  &:hover {
    background-color: #5bc8dc;
  }
}

.active {
  background-color: #f6d918;
}
