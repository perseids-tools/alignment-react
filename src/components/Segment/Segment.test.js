import React from 'react';
import renderer from 'react-test-renderer';

import Segment from './Segment';

const sentence = {
  $: { n: '1' },
  wds: [
    {
      $: { lnum: 'L1' },
      comment: [{ $: { class: 'uri' } }],
      w: [
        { $: { n: '1-1' }, text: [{ _: 'A' }], refs: [{ $: { nrefs: '1-1' } }] },
        { $: { n: '1-2' }, text: [{ _: 'BC' }], refs: [{ $: { nrefs: '1-2 1-3' } }] },
        {
          $: { n: '1-3' },
          text: [{ _: 'D' }],
          comment: [{ _: 'Test', $: { class: 'mark' } }],
          refs: [{ $: { nrefs: '1-4' } }],
        },
        { $: { n: '1-4' }, text: [{ _: 'E' }], refs: [{ $: { nrefs: '1-4' } }] },
        { $: { n: '1-5' }, text: [{ _: 'F' }], refs: [{ $: { nrefs: '1-5 1-6' } }] },
        { $: { n: '1-6' }, text: [{ _: 'G' }], refs: [{ $: { nrefs: '1-5 1-6' } }] },
      ],
    },
    {
      $: { lnum: 'L2' },
      comment: [{ $: { class: 'uri' } }],
      w: [
        { $: { n: '1-1' }, text: [{ _: 'A' }], refs: [{ $: { nrefs: '1-1' } }] },
        { $: { n: '1-2' }, text: [{ _: 'B' }], refs: [{ $: { nrefs: '1-2' } }] },
        { $: { n: '1-3' }, text: [{ _: 'C' }], refs: [{ $: { nrefs: '1-2' } }] },
        { $: { n: '1-4' }, text: [{ _: 'DE' }], refs: [{ $: { nrefs: '1-3 1-4' } }] },
        { $: { n: '1-5' }, text: [{ _: 'F' }], refs: [{ $: { nrefs: '1-5 1-6' } }] },
        { $: { n: '1-6' }, text: [{ _: 'G' }], refs: [{ $: { nrefs: '1-5 1-6' } }] },
      ],
    },
  ],
};

it('renders a sentence', () => {
  const component = (
    <Segment lnum="L1" sentence={sentence} />
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});

it('displays active words differently', () => {
  const active = {
    L1: new Set(['1-2']),
  };

  const component = (
    <Segment lnum="L1" sentence={sentence} active={active} />
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});
