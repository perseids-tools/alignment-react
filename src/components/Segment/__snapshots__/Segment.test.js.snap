// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays active words differently 1`] = `
<div
  className="text"
>
  <span
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    A
  </span>
   
  <span
    className="word active"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    BC
  </span>
   
  <span
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    D
  </span>
   
  <span
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    E
  </span>
   
  <span
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    F
  </span>
   
  <span
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    G
  </span>
</div>
`;

exports[`renders a sentence 1`] = `
<div
  className="text"
>
  <span
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    A
  </span>
   
  <span
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    BC
  </span>
   
  <span
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    D
  </span>
   
  <span
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    E
  </span>
   
  <span
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    F
  </span>
   
  <span
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    G
  </span>
</div>
`;
